## Objective

The objective of this notebook is to explore and understand the Online Retail dataset before any cleaning or preprocessing is performed.

This notebook focuses on:

- Loading the raw dataset.
- Understanding its structure and schema.
- Examining the variables and their data types.
- Creating a data dictionary.
- Assessing data quality by identifying missing values, duplicate records, cancelled transactions, invalid values, and other inconsistencies.
- Producing findings that will guide the Data Preparation phase.

In [1]:
#import libraries

import pandas as pd
import numpy as np

import matplotlib.pyplot as plt
import seaborn as sns


## 2. Load the Dataset

In [2]:
df = pd.read_excel("../data/raw/Online Retail.xlsx")

df.head()

,InvoiceNo,StockCode,Description,Quantity,InvoiceDate,UnitPrice,CustomerID,Country
0,536365,85123A,WHITE HANGING HEART T-LIGHT HOLDER,6,2010-12-01 08:26:00,2.55,17850.0,United Kingdom
1,536365,71053,WHITE METAL LANTERN,6,2010-12-01 08:26:00,3.39,17850.0,United Kingdom
2,536365,84406B,CREAM CUPID HEARTS COAT HANGER,8,2010-12-01 08:26:00,2.75,17850.0,United Kingdom
3,536365,84029G,KNITTED UNION FLAG HOT WATER BOTTLE,6,2010-12-01 08:26:00,3.39,17850.0,United Kingdom
4,536365,84029E,RED WOOLLY HOTTIE WHITE HEART.,6,2010-12-01 08:26:00,3.39,17850.0,United Kingdom


### Interpretation

The first five rows provide an initial preview of the dataset.

Each row represents a product included within a customer invoice rather than a complete order. This means that a single invoice may appear multiple times if multiple products were purchased in the same transaction.

The preview also confirms that the dataset contains information about invoices, products, quantities, prices, customers, and countries.

## 3. Dataset Dimensions


The dimensions of the dataset indicate the number of observations (rows) and variables (columns).

Understanding the dataset size helps estimate its complexity and provides context for subsequent analysis.

In [5]:
rows, columns = df.shape
print(f'Rows: {rows}')
print(f'Columns: {columns}')

Rows: 541909
Columns: 8


### Interpretation

The dataset contains **541,909 transaction records** distributed across **8 variables**.

The large number of observations suggests that the dataset captures a substantial volume of retail transactions, making it suitable for customer segmentation and purchasing behaviour analysis.

The relatively small number of variables also indicates that each attribute is likely to play an important role in understanding customer transactions.

## 4. Dataset Schema

## 4. Dataset Schema

The schema provides an overview of the dataset's structure by displaying the data type of each variable, the number of non-null values, and memory usage.

Understanding the schema helps identify variables that may require preprocessing in later stages.

In [6]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 541909 entries, 0 to 541908
Data columns (total 8 columns):
 #   Column       Non-Null Count   Dtype         
---  ------       --------------   -----         
 0   InvoiceNo    541909 non-null  object        
 1   StockCode    541909 non-null  object        
 2   Description  540455 non-null  object        
 3   Quantity     541909 non-null  int64         
 4   InvoiceDate  541909 non-null  datetime64[ns]
 5   UnitPrice    541909 non-null  float64       
 6   CustomerID   406829 non-null  float64       
 7   Country      541909 non-null  object        
dtypes: datetime64[ns](1), float64(2), int64(1), object(4)
memory usage: 33.1+ MB


### Interpretation

The schema reveals that the dataset contains a combination of numerical, categorical, and datetime variables.

Key observations include:

- **InvoiceNo**, **StockCode**, **Description**, and **Country** are stored as categorical (object) variables.
- **Quantity** is stored as an integer representing the number of products purchased.
- **UnitPrice** is stored as a floating-point number representing the price of each product.
- **InvoiceDate** is stored as a datetime variable, making it suitable for time-based analyses.
- **CustomerID** is stored as a numeric variable but contains missing values, indicating that not every transaction is associated with a registered customer.


## 5. Summary Statistics

Descriptive statistics provide a numerical summary of the dataset and help identify unusual values that may indicate data quality issues.

The statistics include measures such as the mean, minimum, maximum, and standard deviation for numerical variables.

In [7]:
df.describe()

df.describe(include='object')

,InvoiceNo,StockCode,Description,Country
count,541909,541909,540455,541909
unique,25900,4070,4223,38
top,573585,85123A,WHITE HANGING HEART T-LIGHT HOLDER,United Kingdom
freq,1114,2313,2369,495478


## 6. Missing Values Assessment

Missing values reduce the completeness of a dataset and may negatively affect downstream analyses.

This section identifies the number and percentage of missing values present in each variable.

In [8]:
missing_values = df.isnull().sum()

missing_values

InvoiceNo           0
StockCode           0
Description      1454
Quantity            0
InvoiceDate         0
UnitPrice           0
CustomerID     135080
Country             0
dtype: int64

In [9]:
missing_percentage = (missing_values / len(df)) * 100

missing_percentage.sort_values(ascending=False)

CustomerID     24.926694
Description     0.268311
InvoiceNo       0.000000
StockCode       0.000000
Quantity        0.000000
InvoiceDate     0.000000
UnitPrice       0.000000
Country         0.000000
dtype: float64

### Interpretation

The analysis shows that missing values are primarily concentrated in the **CustomerID** column, with a small number of missing values also present in **Description**.

Missing customer identifiers are particularly important because customer segmentation relies on identifying individual customers.


## 7. Duplicate Records

Duplicate records occur when identical observations appear more than once within the dataset.

These records may distort statistical summaries and sales analyses if left untreated.


In [10]:
duplicates = df.duplicated().sum()
print(f'Duplicate rows: {duplicates}')

Duplicate rows: 5268


### Interpretation

The dataset contains duplicate records.

Duplicate transactions may result in the same purchase being counted multiple times, leading to inaccurate calculations of revenue and customer purchasing behaviour.


## 8. Cancelled Transactions


Invoices beginning with the letter **"C"** indicate cancelled customer transactions.

These transactions are identified because they do not represent completed purchases and may influence customer behaviour analysis.

In [11]:
cancelled = df[df["InvoiceNo"].astype(str).str.startswith("C")]

print(f"Cancelled transactions: {len(cancelled)}")

cancelled.head()

Cancelled transactions: 9288


,InvoiceNo,StockCode,Description,Quantity,InvoiceDate,UnitPrice,CustomerID,Country
141,C536379,D,Discount,-1,2010-12-01 09:41:00,27.50,14527.0,United Kingdom
154,C536383,35004C,SET OF 3 COLOURED FLYING DUCKS,-1,2010-12-01 09:49:00,4.65,15311.0,United Kingdom
235,C536391,22556,PLASTERS IN TIN CIRCUS PARADE,-12,2010-12-01 10:24:00,1.65,17548.0,United Kingdom
236,C536391,21984,PACK OF 12 PINK PAISLEY TISSUES,-24,2010-12-01 10:24:00,0.29,17548.0,United Kingdom
237,C536391,21983,PACK OF 12 BLUE PAISLEY TISSUES,-24,2010-12-01 10:24:00,0.29,17548.0,United Kingdom


### Interpretation

The dataset contains cancelled transactions identified by invoice numbers beginning with **"C"**.

These records generally represent product returns or cancelled purchases and should not contribute to analyses of completed customer purchases.

## 9. Negative Quantities

Negative quantities generally represent returned products, stock corrections, or cancelled purchases rather than completed sales.

This section identifies all records containing negative quantities.

In [12]:
negative_quantity = df[df["Quantity"] < 0]

print(f"Negative quantity records: {len(negative_quantity)}")

negative_quantity.head()

Negative quantity records: 10624


,InvoiceNo,StockCode,Description,Quantity,InvoiceDate,UnitPrice,CustomerID,Country
141,C536379,D,Discount,-1,2010-12-01 09:41:00,27.50,14527.0,United Kingdom
154,C536383,35004C,SET OF 3 COLOURED FLYING DUCKS,-1,2010-12-01 09:49:00,4.65,15311.0,United Kingdom
235,C536391,22556,PLASTERS IN TIN CIRCUS PARADE,-12,2010-12-01 10:24:00,1.65,17548.0,United Kingdom
236,C536391,21984,PACK OF 12 PINK PAISLEY TISSUES,-24,2010-12-01 10:24:00,0.29,17548.0,United Kingdom
237,C536391,21983,PACK OF 12 BLUE PAISLEY TISSUES,-24,2010-12-01 10:24:00,0.29,17548.0,United Kingdom


### Interpretation

Negative quantities are present within the dataset.

These observations typically correspond to returned products or inventory adjustments rather than customer purchases.


## 10. Negative Unit Prices

Unit prices are expected to be positive.

This section investigates whether any transactions contain negative pricing values that may indicate data entry errors or exceptional accounting adjustments.

In [13]:
negative_price = df[df["UnitPrice"] < 0]

print(f"Negative unit prices: {len(negative_price)}")

negative_price.head()

Negative unit prices: 2


,InvoiceNo,StockCode,Description,Quantity,InvoiceDate,UnitPrice,CustomerID,Country
299983,A563186,B,Adjust bad debt,1,2011-08-12 14:51:00,-11062.06,NaN,United Kingdom
299984,A563187,B,Adjust bad debt,1,2011-08-12 14:52:00,-11062.06,NaN,United Kingdom


### Interpretation

The analysis checks whether any transactions contain negative unit prices.

If such records exist, they may indicate pricing errors or accounting adjustments and should be investigated during the Data Preparation phase.

If no negative prices are found, this suggests that the pricing information is generally consistent.

## 11. conclusion

The Data Understanding phase provided a comprehensive overview of the Online Retail dataset and identified several data quality issues that require attention before analysis.

The investigation revealed missing customer identifiers, duplicate records, cancelled transactions, and negative quantities, all of which have the potential to affect customer segmentation and sales analysis.
